src/hook.rs

```diff
@@ -0,0 +1,175 @@
+    /// When Lua returns from a function.
+    pub on_returns: bool,
+    /// Before executing a new line, or returning from a function call.
+    pub every_line: bool,
+    /// After a certain amount of instructions. When set to `Some(count)`, `count` is the number of
+    /// instructions to execute before calling the hook.
+    ///
+    /// # Performance
+    ///
+    /// Setting this to a low value will certainly result in a large overhead due to the crate's
+    /// safety layer and convenience wrapped over Lua's low-level hooks. `1` is such an example of
+    /// a high overhead choice.
+    ///
+    /// Please find a number that is high enough so it's not that bad of an issue, while still
+    /// having enough precision for your needs. Keep in mind instructions are additions, calls to
+    /// functions, assignments to variables, etc.; they are very short.
+    pub every_nth_instruction: Option<u32>,
+}
+
+impl HookTriggers {
+    /// Compute the mask to pass to `lua_sethook`.
+    pub(crate) fn mask(&self) -> c_int {
+        let mut mask: c_int = 0;
+        if self.on_calls { mask |= ffi::LUA_MASKCALL }
+        if self.on_returns { mask |= ffi::LUA_MASKRET }
+        if self.every_line { mask |= ffi::LUA_MASKLINE }
+        if self.every_nth_instruction.is_some() { mask |= ffi::LUA_MASKCOUNT }
+        mask
+    }
+
+    /// Returns the `count` parameter to pass to `lua_sethook`, if applicable. Otherwise, zero is
+    /// returned.
+    pub(crate) fn count(&self) -> c_int {
+        self.every_nth_instruction.unwrap_or(0) as c_int
+    }
+}
+
+impl Default for HookTriggers {
+    fn default() -> Self {
+        HookTriggers {
+            on_calls: false,
+            on_returns: false,
+            every_line: false,
+            every_nth_instruction: None
+        }
+    }
+}
+
+/// This callback is passed to `lua_sethook` and gets called whenever debug information is received.
+pub(crate) unsafe extern "C" fn hook_proc(state: *mut lua_State, ar: *mut lua_Debug) {
+    callback_error(state, || {
+        if ffi::lua_getinfo(state, cstr!("nSltu"), ar) == 0 {
+            rlua_panic!("lua_getinfo failed")
+        }
+
+        let lua = Lua::from_state(state);
+        let debug = Debug {
+            name: ptr_to_str((*ar).name as *const i8),
+            namewhat: ptr_to_str((*ar).namewhat as *const i8),
+            what: ptr_to_str((*ar).what as *const i8),
+            source: ptr_to_str((*ar).source as *const i8),
+            curr_line: (*ar).currentline as u32,
+            line_defined: (*ar).linedefined as u32,
+            last_line_defined: (*ar).lastlinedefined as u32,
+            num_ups: (*ar).nups as u32,
+            num_params: (*ar).nparams as u32,
+            is_vararg: (*ar).isvararg == 1,
+            is_tailcall: (*ar).istailcall == 1,
+            short_src: str::from_utf8(mem::transmute((*ar).short_src.as_ref()))
+                .and_then(|r| Ok(Some(Cow::from(r))))
+                .unwrap_or(None),
```

Comment:
Please don't use a transmute here, please do `(*ar).short_src as *const u8` instead.

Reply:
I think it was because it was a fixed list, not sure if the cast was allowed

Reply:
`(*ar).short_src` has type `*const i8`, right?  That's why you're transmuting?

Reply:
Ohhhh, no it's an array, hold on let me think about this some more, I still don't like `mem::transmute`, *especially* without the types being specified

Reply:
Okay, I think it would be cleaner to do slice::from_raw_parts like is done in `ptr_to_str` again, and possibly use the same function that the rest of the conversion does.  As a general rule, don't call `mem::transmute` without type annotation, even if you know the types at the time.  It's incredibly dangerous in the presence of refactoring, because if you refactor and change the types `mem::transmute` will happily produce any type you desire that is the same size.

At the very least it should specify the types with turbofish.